File: java/src/readerswriters/Book.java
package readerswriters;

import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

class Book {

    private int maxReaders;
    private boolean wakeUpAllReaders;

    private int readersNumber;
    private boolean isWriting = false;

    private int readersInARowCounter;

    private ReentrantLock lock = new ReentrantLock();
    private Condition writing = lock.newCondition();
    private Condition reading = lock.newCondition();

    public Book(int maxReaders, boolean wakeUpAllReaders) {
        super();
        this.maxReaders = maxReaders;
        this.wakeUpAllReaders = wakeUpAllReaders;
        this.readersInARowCounter = 0;
    }

    public void enterWriting(){
        lock.lock();

        while(isWriting == true || readersNumber > 0){
            try{
                writing.await();
            }
            catch (InterruptedException e){
                e.printStackTrace();
            }
        }
        isWriting = true;

        // System.out.println(String.format("Writer-entry: %s is about to enter...", Thread.currentThread().getName()));

        lock.unlock();
    }

    /**
     * Writer leaves its critical section.
     * @returns The number of readers entered in a row before a writer got the resource.
     */
    public int leaveWriting(){
        lock.lock();

        isWriting = false;
        int tmpCounter = readersInARowCounter;
        readersInARowCounter = 0;

        if (wakeUpAllReaders) {
            reading.signalAll();
        } else {
            reading.signal();
        }
        writing.signal();

        // System.out.println(String.format("Writer-exit: %s is about to leave...", Thread.currentThread().getName()));

        lock.unlock();
        return tmpCounter;
    }

    public void enterReading(){
        lock.lock();

        while(isWriting == true || readersNumber >= maxReaders){
            try{
                reading.await();
            }
            catch (InterruptedException e){
                e.printStackTrace();
            }
        }
        readersNumber++;
        readersInARowCounter++;
        // System.out.println(String.format("Reader-entry: %s is about to enter...", Thread.currentThread().getName()));

        lock.unlock();
    }

    public void leaveReading(){
        lock.lock();

        readersNumber--;
        if (wakeUpAllReaders) {
            reading.signalAll();
        } else {
            reading.signal();
        }
        writing.signal();

        // System.out.println(String.format("Reader-exit: %s is about to leave...", Thread.currentThread().getName()));

        lock.unlock();
    }

    /**
     * Writer's critical section.
     * @returns The number of readers entered in a row before a writer got the resource.
     */
    public int write() {
        enterWriting();
        return leaveWriting();
    }

    public void read() {
        enterReading();
        leaveReading();
    }
}

File: dotnet/ReadersWritersDotnet/ReadersWritersDotnet/Book.cs
//using System;
using System.Threading;

namespace ReadersWritersDotnet
{
    public class Book : IBook
    {
        private readonly int _maxReaders;
        private readonly bool _wakeUpAllReaders;
        private readonly object _reading;
        private readonly object _writing;

        private int _readersNumber;
        private int _readersInARowCounter;

        public Book(int maxReaders, bool wakeUpAllReaders) {
            _maxReaders = maxReaders;
            _wakeUpAllReaders = wakeUpAllReaders;
            _reading = new object();
            _writing = new object();

            _readersNumber = 0;
            _readersInARowCounter = 0;
        }

        /// <summary>Writer's critical section.</summary>
        /// <returns>The number of readers entered in a row before a writer got the resource.</returns>
        public int Write(){
            //Console.WriteLine($"{Thread.CurrentThread.Name} (writer) - step in, before lock...");
            int tmpCounter;
            lock (_writing)
            {
                lock (_reading)
                {
                    while(_readersNumber > 0)
                    {
                        Monitor.Wait(_reading);
                    }
                    
                    tmpCounter = _readersInARowCounter;
                    _readersInARowCounter = 0;
                    
                    if (_wakeUpAllReaders)
                    {
                        Monitor.PulseAll(_reading);
                    }
                    else Monitor.Pulse(_reading);
                }
            }
            //Console.WriteLine($"{Thread.CurrentThread.Name} (writer) - finished.");
            return tmpCounter;
        }

        private void EnterReading(){
            //Console.WriteLine($"{Thread.CurrentThread.Name} (reader) - step in, before lock...");
            lock (_reading)
            {
                while(_readersNumber >= _maxReaders)
                {
                    Monitor.Wait(_reading);
                }
                _readersNumber++;
                _readersInARowCounter++;
            }
            //Console.WriteLine($"{Thread.CurrentThread.Name} (reader) - is in...");
        }

        private void LeaveReading(){
            //Console.WriteLine($"{Thread.CurrentThread.Name} (reader) - leaving, before lock...");
            lock (_reading)
            {
                _readersNumber--;
                if (_wakeUpAllReaders)
                {
                    Monitor.PulseAll(_reading);
                }
                else Monitor.Pulse(_reading);
            }
            //Console.WriteLine($"{Thread.CurrentThread.Name} (reader) - finished.");
        }

        public void Read() {
            EnterReading();
            LeaveReading();
        }
    }
}